### PR TITLE
fix 2871

### PR DESCRIPTION
**Pull request by dsyme**
This simplifies the type traversal code prior to calling ``CreateType`` and in the process fixes https://github.com/Microsoft/visualfsharp/issues/2871 by ensuring the enclosing types of a type are created before the type itself.  The ``priority``parameter served no useful purpose since enclosing types always need to be created first and was part of the cause of the problem.

**Please don't pull this until we have also run the Mono tests that are in fsharp/fsharp** as Mono versions are sensitive to order of CreateType calls (they have been fixing issues in this area in Mono 5.0)

### CODE REVIEW

File: src/absil/ilreflect.fs [hunk 1]
```diff
@@ -1845,78 +1858,100 @@ let rec buildTypeDefPass3 cenv nesting modB emEnv (tdef : ILTypeDef) =
...
 let getEnclosingTypeRefs (tref:ILTypeRef) = 
    match tref.Enclosing with 
    | [] -> []
    | h :: t -> List.scan (fun tr nm -> mkILTyRefInTyRef (tr,nm)) (mkILTyRef(tref.Scope, h)) t
 
-let rec getTypeRefsInType valueTypesOnly typ acc = 
+[<RequireQualifiedAccess>]
+type AllTypes = Yes | No
+
+// Find all constitunt type references
```
Comment (brodyberg) [inline src/absil/ilreflect.fs]: spelling

**Comment by dsyme** [comment]
OK, I ran this through http://github.com/fsharp/fsharp and all was OK.  This is now ready for merge after review

File: src/absil/ilreflect.fs [hunk 2]
```diff
@@ -1845,78 +1858,100 @@ let rec buildTypeDefPass3 cenv nesting modB emEnv (tdef : ILTypeDef) =
...
-    | ILType.Array (_,eltType) -> if valueTypesOnly then acc else getTypeRefsInType valueTypesOnly eltType acc
-    | ILType.Value tspec -> tspec.TypeRef :: List.foldBack (getTypeRefsInType valueTypesOnly) tspec.GenericArgs acc
-    | ILType.Boxed tspec -> if valueTypesOnly then acc else tspec.TypeRef :: List.foldBack (getTypeRefsInType valueTypesOnly) tspec.GenericArgs acc
+    | ILType.Void 
+    | ILType.TypeVar _  -> acc
+    | ILType.Ptr eltType | ILType.Byref eltType -> 
+        getTypeRefsInType allTypes eltType acc
+    | ILType.Array (_,eltType) -> 
+        match allTypes with 
+        | AllTypes.No -> acc
```
Comment (KevinRansom) [inline src/absil/ilreflect.fs]: Before it was clear that it applied to ValueTypesOnly from the naming.  It's not clear what distinguishes between alltypes yes and alltypes no?

**Comment by dsyme** [comment]
@KevinRansom I clarified the code, thanks

**Pull request merged**